### ISSUE TITLE
Replace print statements by warnings or stderr in NCBIXML

### ISSUE THREAD

**Issue by JoaoRodrigues** [biopython/biopython#4262]
Follow-up to #4256, we should replace the print statements in the NCBIXML module by either calls to `warnings` or to `sys.stderr`, to be easier to ignore/discard.

Labelling this as a good first issue, but feel free to remove the label @peterjc.

**Comment by BabaYaga1221** [comment]
I am interested to work on this issue and I am currently active as I also create a pull request recently. Would you like to assign this issue to me?  Thanks @JoaoRodrigues 
~Regards
Farhan

**Comment by peterjc** [comment]
Actually looking at this fresh:

https://github.com/biopython/biopython/blob/master/Bio/Blast/NCBIXML.py

We have five old usages of ``print(...)`` which are all conditional on debug mode. You might argue those should be replaced with ``sys.stderr.write(...)`` but that seems fine. Either way, these are user requested output.

Then we have one new usage of ``print(...)`` for a warning added in #4256. This is not user requested output, but a warning that something wasn't right in the XML. This I think should be done with something like this:

```python
import warnings
from Bio import BiopythonParserWarning

...

warnings.warn(f"NCBIXML: Ignored: {self._value!r}", BiopythonParserWarning)
```

**Comment by mdehoon** [comment]
> Actually looking at this fresh:
> 
> https://github.com/biopython/biopython/blob/master/Bio/Blast/NCBIXML.py
> 
> We have five old usages of `print(...)` which are all conditional on debug mode. You might argue those should be replaced with `sys.stderr.write(...)` but that seems fine. Either way, these are user requested output.
> 
> Then we have one new usage of `print(...)` for a warning added in #4256. This is not user requested output, but a warning that something wasn't right in the XML. This I think should be done with something like this:
> 
> ```python
> import warnings
> from Bio import BiopythonParserWarning
> 
> ...
> 
> warnings.warn(f"NCBIXML: Ignored: {self._value!r}", BiopythonParserWarning)
> ```

By default, warnings get printed only once, even if triggered multiple times, and are easily missed. I would suggest to raise an Exception by default, and use the `debug` argument to let the user ignore exceptions.

**Comment by mdehoon** [comment]
@BabaYaga1221  What do you mean by twink?

**Comment by JoaoRodrigues** [comment]
For the message we just added in #4256 I'd keep it as a message, since using debug will generate a _lot_ of other output. Ideally, we'd use `logging` and set some messages (the real debug ones) to `log.debug` and the ones like in #4256 to `log.warning`, but that might be too much of an overhaul?

**Comment by BabaYaga1221** [comment]
Actually, I want to say that if we add a flag argument whenever a user tries to trigger a warning, might solve this issue when the user triggered a warning multiple times. I think this might be a solution but still, this is not an optimized way to handle this. @mdehoon

**Comment by peterjc** [comment]
Twink was probably an autocorrection/typo for tweak, as in change slightly.

Personally I would use the warnings module for the newly added message, second choice writing it to stderr.

For the existing debug messages, I am inclined not to change anything.

Joao's suggestion to use the logging module would fit with an overhaul of other debug usage in Biopython, including the PDB parser. It doesn't seem sensible to adopt this just for this one new message.

**Comment by mdehoon** [comment]
We should decide how to handle these kinds of warnings in general in Biopython. But we don't need to update all of Biopython at the same time to whatever approach we decide on; this can happen gradually over time.

**Comment by claudiofr** [comment]
It seems that @BabaYaga1221 wanted to work on this issue but I haven't seen any updates since Mar 17. Is anyone actively working on this issue? If not would you still consider this a good first issue for someone who is contributing to this project for the first time?

**Comment by tulga-rdn** [comment]
It also seems that the print statements are constructed inconsistently. In some print statements, the output is constructed by concatenating strings, whereas in some other statements, the output is constructed by using f-strings. If the issue is still open, I can work on this. @peterjc @mdehoon

**Comment by mdehoon** [comment]
@tulga-rdn  Bio.Blast.NCBIXML is likely to be deprecated and replaced by a new parser some time after release 1.82 is out.

**Comment by uxritu** [comment]
Hi, I'm new to open-source contributions. Can I work on this issue

**Comment by uxritu** [comment]
> [@uxritu](https://github.com/uxritu) As per the previous comment this is low priority, but equally low risk if you wanted to try. We would expect changes to the test suite to check for and/or silence the warnings when running the tests.
> 
> --
> 
> Note [@uxritu](https://github.com/uxritu) has written the same comment on multiple issues today - which I am initially willing to assuming is merely over enthusiasm.

yes new learning could be over enthusiasm to you ..from wet lab to dry lab is new and tough for me ..so i am trying my hands

**Comment by rakirs2** [comment]
@uxritu  are you still working on this?

**Comment by head-ease08** [comment]
Made a replace, feel free to review!

**Comment by kirandharmar867-maker** [comment]
Hello @JoaoRodrigues 👋  
I’d like to work on this issue.  
Could you please confirm if it’s still open and if there are any specific guidelines or tests you’d like me to follow?

Thanks!

**Comment by peterjc** [comment]
@kirandharmar867-maker please read #5030, see if you understand where that stalled, and perhaps try to finish it as a new PR.

**Comment by kirandharmar867-maker** [comment]
> [@kirandharmar867-maker](https://github.com/kirandharmar867-maker) please read [#5030](https://github.com/biopython/biopython/pull/5030), see if you understand where that stalled, and perhaps try to finish it as a new PR.

~Sorry for the oversight,  I hadn’t checked #5030 earlier.

I’ve gone through it now and I understand where the work stalled. 
I’LL try to continue from there and work on finishing it as a new PR.

Thanks for the guidance!
@peterjc
